File: frontend/src/types/product.ts
import { Brand } from './brand';
import { Category } from './category';
import { Color } from './color';
import { ProductSpecification } from './specDefinition';

export interface ProductImage {
  id: number;
  imageUrl: string;
  imagePublicId?: string;
  displayOrder: number;
}

export interface ProductVariant {
  id: number;
  productId: number;
  color: Color;
  sku: string;
  additionalPrice: number;
  stockQuantity: number;
  variantImage: string;
  variantImagePublicId?: string;
  isActive: boolean;
}

export interface ProductVariantCreateRequest {
  colorId: number;
  sku: string;
  additionalPrice: number;
  stockQuantity: number;
  variantImage: string;
  variantImagePublicId?: string;
}

export interface ProductVariantUpdateRequest {
  colorId?: number;
  sku?: string;
  additionalPrice?: number;
  stockQuantity?: number;
  variantImage?: string;
  variantImagePublicId?: string;
  isActive?: boolean;
}

export interface Product {
  id: number;
  productName: string;
  slug: string;
  brand: Brand;
  category: Category;
  summary: string;
  description: string;
  basePrice: number;
  discountPrice: number;
  thumbnail: string;
  thumbnailPublicId: string;
  weightGram?: number;
  lengthCm?: number;
  widthCm?: number;
  heightCm?: number;
  isFeatured: boolean;
  isActive: boolean;
  createdAt: string;
  images?: ProductImage[];
  variants?: ProductVariant[];
  specifications?: ProductSpecification[];
}

File: backend/src/main/java/com/company/electricfanshop/entity/product/Product.java
package com.company.electricfanshop.entity.product;

import com.company.electricfanshop.entity.review.Review;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "products")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    private String productName;

    private String slug;

    // Many Products -> One Brand
    @ManyToOne
    @JoinColumn(name = "brand_id")
    private Brand brand;

    // Many Products -> One Category
    @ManyToOne
    @JoinColumn(name = "category_id")
    private Category category;

    private String summary;

    private String description;

    private BigDecimal basePrice;

    private BigDecimal discountPrice;
    private String thumbnail;
    private String thumbnailPublicId;

    private Integer engineCapacity;

    private Integer weightGram;
    private Integer lengthCm;
    private Integer widthCm;
    private Integer heightCm;

    @Builder.Default
    private Boolean isFeatured = false;

    @Builder.Default
    private Boolean isActive = true;

    @CreationTimestamp
    private LocalDateTime createdAt;

    // Relationships
    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL)
    private List<ProductImage> images;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL)
    private List<ProductVariant> variants;

    @OneToMany(mappedBy = "product")
    private List<Review> reviews;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL)
    private List<ProductSpecification> specifications;
}